DOC Document code moved from cms to framework (#625)
silverstripe/developer-docs: en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md

In the above screenshot, "MODIFIED" and "EN_US" are status flags. The "MODIFIED" flag is provided by the `Versioned` extension, and the "EN_US" flag is provided by [`tractorcow/silverstripe-fluent`](https://github.com/tractorcow-farm/silverstripe-fluent/).

Note that in the site tree only one status flag will be displayed when you're editing a record, due to the limited space available.

Some status flags have additional context which can be viewed by hovering over them with your mouse cursor.

## Adding and removing status flags

You can easily add and remove status flags.

The customization of status flags can be done either directly in the model class by overriding the [`ModelData::getStatusFlags()`](api:SilverStripe\Model\ModelData::getStatusFlags()) method, or by [applying an extension](/developer_guides/extending/extensions/) that implements the `updateStatusFlags` extension hook.

Add status flags by adding new items to the `$flags` array, and remove existing flags by unsetting the relevant key in the array.

### Directly on the model

```php
namespace App\Model;

use SilverStripe\ORM\DataObject;

class MyModel extends DataObject
{
    // ...

>     private static array $db = [
>         'IsPublishScheduled' => 'Boolean',
>     ];
> 
>     public function getStatusFlags(bool $cached = true): array
>     {
>         $this->beforeExtending('updateStatusFlags', function (array &$flags) {
>             if ($this->IsPublishScheduled) {
>                 $flags['scheduledtopublish'] = _t(__CLASS__ . 'SCHEDULE_PUBLISH', 'Scheduled To publish');

Reviewer: ```suggestion
    private static array $db = [
        'MyColour' => 'Varchar',
    ];

    public function getStatusFlags(bool $cached = true): array
    {
        $this->beforeExtending('updateStatusFlags', function (array &$flags) {
            if (in_array($this->MyColour, ['Red', 'Green'])) {
                $flags['mycolour'] = $this->MyColour;
```

I think it's better to just use arbitary example data rather than using "IsPublishScheduled" which kind of implies there's "scheduled publishing" functionality somewhere in the CMS, and I think we should just avoid "publishing" in general here as that's related to Versioning, and the point of this is we're abstracting out something that's traditionally been there mostly just there to provide versioned status badges

Also I don't think including the _t() function here is helpful as that's another concept all together, and I think the example should only be demonstrating the primary things it's trying to demonstrate

The simpler the example is the easier it is to follow

Reviewer 2: As discussed in person since the problem you had was the word "published" because it's similar to versioning concepts I've just reworked it to avoid that word while still giving a meaningful example that hints at the sorts of use cases this might have.

I think using `_t()` in examples promotes best practice but as discussed will leave it out for the sake of getting this merged.

I've also updated the screenshot so it reflects the custom flag from the example.

            }
        });
        return parent::getStatusFlags($cached);
    }
}
```

### Via an extension

```php
namespace App\Extension;

use SilverStripe\Core\Extension;

class SchedulePublishExtension extends Extension
{
    private static array $db = [
        'IsPublishScheduled' => 'Boolean',
    ];

    protected function updateStatusFlags(array &$flags): void
    {
        if ($this->getOwner()->IsPublishScheduled) {
            $flags['scheduledtopublish'] = _t(__CLASS__ . 'SCHEDULE_PUBLISH', 'Scheduled To publish');
        }
    }
}
```

### Including additional context

Sometimes the text of a status flag won't convey all the information you want it to, since it is so short.

You can include additional context which will be displayed when hovering over your status flags. This is done using an associative array where the main text of the flag is in the `text` key, and the additional information is in the `title` key:

```php

> namespace App\Model;

Reviewer: I think this example should be changed like I've recommended above i.e. just use "MyColour" and don't use _t()

Reviewer 2: Done
